### DISCUSSION TITLE
How do I use the request decoder to access values of the HTTP Request that are not in the header nor the body?

### DISCUSSION THREAD

**Discussion by ayx-orion** [swaggest/rest#175] [Q&A] [answered]
I'm trying to access some values in the HTTP request header (specifically the value of "remoteIP" )

i've created an Input struct that the request decoder is populating but i couldn't find any examples of getting values that aren't in the path, param, queryString, json body, etc. 

my http request looks like this (from server log, doesn't show json payload values)
```
httpRequest=
{
  "header": {
    "accept":"*/*",
    "accept-encoding": "gzip, deflate, br",
    "connection":"keep-alive",
    "content-length":"280",
    "content-type":"application/json",
    "postman-token":"019a95a5-9a0c-4e14-8b4e-1e529957f668",
    "user-agent":"PostmanRuntime/7.29.0"
  },
  "proto":"HTTP/1.1",
  "remoteIP":"172.20.176.1:56912",
  "requestID":"orion-hyper-vm/5LgNIbnAgS-000001",
  "requestMethod":"POST",
  "requestPath":"/api/v1/archives/verification",
  "requestURL":"http://172.20.176.126:31337/api/v1/archives/verification",
  "scheme":"http"
}
```
as you can see, the `remoteIP` (along with some other values) is not part of the `headers` dictionary, but is still part of the HTTP request itself. how can i access those values?

this is my struct that i'm using for the usecase Input:
```
type VerificationRequest struct {
	ContentType    string `header:"content-type"`   // this works, is correctly initiated with the value "application/json"
	XForwardedHost string `header:"remoteIP"`   // <-- this doesn't work, kind of expected, but not sure what else to do here :( 
	EntityId       string `json:"entityId" pattern:"^[0-7][0-9A-HJKMNP-TV-Z]{25}$"`
	Uri            string `json:"uri"`
	CallbackUri    string `json:"callbackUri"`
}
```

EDIT: I think the value of "remoteIP" in that request _should_ correspond to the struct member `http.Request.RemoteAddr`

**Comment by vearutop** [comment, marked as answer]
Currently, the easiest way to access remote addr is to embed `request.EmbeddedSetter` and then get direct access to `*http.Request`.

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"

	"github.com/swaggest/openapi-go/openapi31"
	"github.com/swaggest/rest/request"
	"github.com/swaggest/rest/web"
	swgui "github.com/swaggest/swgui/v5cdn"
	"github.com/swaggest/usecase"
)

func main() {
	s := web.NewService(openapi31.NewReflector())

	type inputWithReq struct {
		request.EmbeddedSetter

		Locale string `query:"locale" default:"en-US" pattern:"^[a-z]{2}-[A-Z]{2}$" enum:"ru-RU,en-US"`
		Name   string `path:"name" minLength:"3" default:"Jane"`
	}

	u := usecase.NewInteractor(func(ctx context.Context, input inputWithReq, output *struct{}) error {
		fmt.Println("Remote addr:", input.Request().RemoteAddr) // <------------------- *http.Request access

		return nil
	})

	s.Get("/hello/{name}", u)

	s.Docs("/docs", swgui.New)
	log.Println("http://localhost:8011/docs")
	if err := http.ListenAndServe("localhost:8011", s); err != nil {
		log.Fatal(err)
	}
}

```

Perhaps, we can introduce some field tag to implement automated mapping of remote addr (and maybe some other fields) into structure fields.

**Reply by ayx-orion** [reply]
Thanks a lot @vearutop, I'll give that a try.

I put down a TODO to implement my own LoadFromHTTPRequest method for the InputStruct I was using but I think this will be much easier and quicker

**Reply by ayx-orion** [reply]
using the embedded request was super easy and "just worked". Thank you again @vearutop
